File: test/test_GRF_disk_r64.f90
! test_GRF_disk_r64.f90
! Pure Fortran double-precision GRF disk test.
! Mirrors matlab/test_GRF_disk_r64.m:
!   - unit-circle geometry from zinit_r64
!   - list1 direct panel sum
!   - list2 far-panel direct + near-panel sdspecialquad_r64
!   - harmonic exact solution u = 0.70*Re(z^8) + 1.00*Im(z^7)

program test_GRF_disk_r64
  use utils_mod
  use geometry_mod
  use specialquad_mod
  implicit none

  integer(8), parameter :: npan=60, p=16, np=npan*p
  integer(8), parameter :: Ng=100, Ngtot=Ng*Ng
  real(r64),  parameter :: dlim=0.8_r64, coef1=0.70_r64, coef2=1.00_r64
  integer(8), parameter :: pow1=8, pow2=7

  ! Panel arrays
  real(r64)    :: sinter(npan+1), sinterdiff(npan), T(p), W(p), Dgl(p,p)
  real(r64)    :: wq(np), awzp(np), ub(np), unb(np), panlen(npan)
  complex(r64) :: z(np), zp(np), zpp(np), nz(np), wzp(np), zinter(npan+1)

  ! Target arrays
  complex(r64) :: zg(Ngtot)
  real(r64)    :: xg(Ng), yg(Ng), u(Ngtot)
  integer(8)   :: mylist(Ngtot), list1(Ngtot), list2(Ngtot), nl1, nl2, nout

  ! Work
  integer(8)   :: k, kk, j, jj, i1, i2
  real(r64)    :: d, err1, err2, unum, unumS, unumK, pi, fx, fy
  complex(r64) :: zt(1), tmp1, tmp2, ci
  real(r64)    :: As(p,1), A1(p,1), A2(p,1), A3(p,1), A4(p,1)
  complex(r64) :: Ad(p,1)
  logical      :: is_out

  pi = acos(-1.0_r64)
  ci = cmplx(0.0_r64, 1.0_r64, r64)

  ! ---- panel setup ----
  call gauss_r64(p, T, W, Dgl)
  do k = 1, npan+1
    sinter(k) = real(k-1, r64) / real(npan, r64)
  end do
  do k = 1, npan
    sinterdiff(k) = 1.0_r64 / real(npan, r64)
  end do
  call zinit_r64(npan, p, sinter, sinterdiff, T, W, z, zp, zpp, nz, wq, wzp, awzp, zinter)

  ! ---- boundary data ----
  do j = 1, np
    ub(j) = coef1*real(z(j)**pow1, r64) + coef2*aimag(z(j)**pow2)
    tmp1  = real(pow1, r64) * z(j)**(pow1-1)
    tmp2  = real(pow2, r64) * z(j)**(pow2-1)
    fx    = coef1*real(tmp1, r64) + coef2*aimag(tmp2)
    fy    = coef1*real(ci*tmp1, r64) + coef2*aimag(ci*tmp2)
    unb(j) = real(nz(j), r64)*fx + aimag(nz(j))*fy
  end do

  ! ---- panel arc lengths ----
  do kk = 1, npan
    i1 = (kk-1)*p + 1
    i2 = kk*p
    panlen(kk) = sum(awzp(i1:i2))
  end do

  ! ---- target grid ----
  do k = 1, Ng
    xg(k) = -1.2_r64 + 2.4_r64*real(k-1, r64)/real(Ng-1, r64)
    yg(k) = xg(k)
  end do
  do k = 1, Ng
    do j = 1, Ng
      zg((k-1)*Ng + j) = cmplx(xg(k), yg(j), r64)
    end do
  end do

  ! ---- classify targets ----
  nl1 = 0
  nl2 = 0
  nout = 0
  do k = 1, Ngtot
    mylist(k) = 0
    is_out = abs(zg(k)) > 1.0_r64
    if (.not. is_out) cycle
    nout = nout + 1
    mylist(k) = 1
    do kk = 1, npan
      i1 = (kk-1)*p + 1
      i2 = kk*p
      d = minval(abs(z(i1:i2) - zg(k))) / panlen(kk)
      if (d < dlim) then
        mylist(k) = 2
        exit
      end if
    end do
    if (mylist(k) == 1) then
      nl1 = nl1 + 1
      list1(nl1) = k
    else
      nl2 = nl2 + 1
      list2(nl2) = k
    end if
  end do

  ! ---- list1: full direct sum ----
  do jj = 1, nl1
    k = list1(jj)
    unum = 0.0_r64
    do j = 1, np
      unumS = -log(abs(z(j)-zg(k))) * awzp(j) / (2.0_r64*pi)
      unumK =  aimag(wzp(j)/(zg(k)-z(j))) / (2.0_r64*pi)
      unum = unum + ub(j)*unumK - unb(j)*unumS
    end do
    u(k) = abs(unum)
  end do

  ! ---- list2: far direct + near special quad ----
  do jj = 1, nl2
    k = list2(jj)
    unum = 0.0_r64
    do kk = 1, npan
      i1 = (kk-1)*p + 1
      i2 = kk*p
      d = minval(abs(z(i1:i2)-zg(k))) / panlen(kk)

      if (d > dlim) then
        do j = i1, i2
          unumS = -log(abs(z(j)-zg(k))) * awzp(j) / (2.0_r64*pi)
          unumK =  aimag(wzp(j)/(zg(k)-z(j))) / (2.0_r64*pi)
          unum = unum + ub(j)*unumK - unb(j)*unumS
        end do
      else
        zt(1) = zg(k)
        call sdspecialquad_r64(1_8, zt, p, z(i1:i2), nz(i1:i2), wzp(i1:i2), &
             zinter(kk), zinter(kk+1), 1_8, As, Ad, A1, A2, A3, A4)
        do j = 1, p
          unum = unum + ub(i1+j-1)*real(Ad(j,1), r64) - unb(i1+j-1)*As(j,1)
        end do
      end if
    end do
    u(k) = abs(unum)
  end do

  ! ---- GRF residual diagnostics (outside only): u should be near zero ----
  err1 = 0.0_r64
  err2 = 0.0_r64
  do jj = 1, nl1
    k = list1(jj)
    err1 = max(err1, u(k))
  end do
  do jj = 1, nl2
    k = list2(jj)
    err2 = max(err2, u(k))
  end do

  write(*,'(A,I0,A,I0,A,I0)') 'Counts: outside=', nout, ', list1=', nl1, ', list2=', nl2
  write(*,'(A,1PE12.4,A,0PF8.2)') 'List1 GRF residual max|u|: ', err1, '  log10=', log10(err1 + epsilon(1.0_r64))
  write(*,'(A,1PE12.4,A,0PF8.2)') 'List2 GRF residual max|u|: ', err2, '  log10=', log10(err2 + epsilon(1.0_r64))

  if (nl1 == 0 .or. nl2 == 0) then
    write(*,*) 'test_GRF_disk_r64: invalid target classification'
    stop 2
  end if

  write(*,*) 'test_GRF_disk_r64: completed'

end program test_GRF_disk_r64

File: test/test_GRF_disk_r64.m
% GRF_disk_r64.m
% Double-precision GRF disk test via poisson2d MEX interface.
% Mirrors GRF_disk.m; uses Fortran for gauss, zinit, lap2d_eval.
% Near-panel special quadrature falls back to MATLAB SDspecialquad (from GRF_disk.m).
%
% Requires: poisson2d_mex built (make mex), SDspecialquad.m in this folder.

testdir = fileparts(mfilename('fullpath'));
addpath(fullfile(testdir, '../matlab'))

p     = 16;
npan  = 60;
dlim  = 0.8;
coef1 = 0.70;  pow1 = 8;
coef2 = 1.00;  pow2 = 7;

% --- panel setup via MEX ---
[T, W, ~] = gauss_mex(p);
sinter     = linspace(0,1,npan+1)';
sinterdiff = ones(npan,1)/npan;
[z, zp, zpp, nz, w, wzp, awzp, zinter] = zinit_mex(npan, p, sinter, sinterdiff, T, W);
% [z2,zp2,~,nz2,~,wzp2,awzp2,zinter2] = zinit(sinter,sinterdiff,T,W,npan);

% target grid
Ngtot = Ng*Ng;
xg = linspace(-1.2,1.2,Ng);
yg = linspace(-1.2,1.2,Ng);
zg = zeros(Ngtot,1);
for k = 1:Ng
  zg((k-1)*Ng+(1:Ng)) = xg(k) + 1i*yg;
end

% boundary data from high-order harmonic test function (ReviewerX_Q4Q5 style)
f  = coef1*real(z.^pow1) + coef2*imag(z.^pow2);
fx = coef1*real(pow1*z.^(pow1-1)) + coef2*imag(pow2*z.^(pow2-1));
fy = coef1*real(1i*pow1*z.^(pow1-1)) + coef2*imag(1i*pow2*z.^(pow2-1));
ub  = f;
unb = real(nz).*fx + imag(nz).*fy;

% classify outside targets into list1/list2
panlen = zeros(npan,1);
for kk = 1:npan
  myind = (kk-1)*p + (1:p);
  panlen(kk) = sum(awzp(myind));
end

inout = Ineval(zg);
mylist = inout;
for k = 1:Ngtot
  if inout(k)
    for kk = 1:npan
      myind = (kk-1)*p + (1:p);
      d = min(abs(z(myind)-zg(k)))/panlen(kk);
      if d < dlim
        mylist(k) = 2;
      end
    end
  end
end
mylist1 = find(mylist==1)';
mylist2 = find(mylist==2)';

%
u = eps*ones(Ngtot,1);

%
disp('List1 starts')
for k = mylist1
  UnumS = -log(abs(z-zg(k))).*awzp/2/pi;
  UnumK = imag(wzp./(zg(k)-z))/2/pi;
  Unum = ub.'*UnumK - unb.'*UnumS;
  u(k) = abs(Unum);
end

disp('List2 starts')
for k = mylist2
  Unum = 0;
  for kk = 1:npan
    myind = (kk-1)*p + (1:p);
    zsc_kk = z(myind);
    wzp_kk = wzp(myind);
    awzp_kk = abs(wzp_kk);
    ub_kk = ub(myind);
    unb_kk = unb(myind);
    d = min(abs(zsc_kk-zg(k)))/panlen(kk);

    if d > dlim
      UnumS = -log(abs(zsc_kk-zg(k))).*awzp_kk/2/pi;
      UnumK = imag(wzp_kk./(zg(k)-zsc_kk))/2/pi;
      Unum = Unum + ub_kk.'*UnumK - unb_kk.'*UnumS;
    else
      a_kk = zinter(kk);
      b_kk = zinter(kk+1);
      nz_kk = nz(myind);

      % tt = struct('x', zg(k));
      % sk = struct('x', zsc_kk, 'nx', nz_kk, 'wxp', wzp_kk);
      % [Sclose, Dclose] = SDspecialquad(tt, sk, a_kk, b_kk, 'e');
      % Unum = Unum + ub_kk.'*real(Dclose)' - unb_kk.'*real(Sclose)';

      [Sclose, Dclose] = sdspecialquad_mex( 1, zg(k), p, zsc_kk, nz_kk, wzp_kk, a_kk, b_kk, 1);
      Unum = Unum + ub_kk.'*real(Dclose(:,1)) - unb_kk.'*Sclose(:,1);
    end
  end
  u(k) = abs(Unum);
end

u(u<eps)=eps;
F1=zeros(Ng);
for k=1:Ng
  F1(1:Ng,k)=log10(u((k-1)*Ng+(1:Ng)));
end
figure(1),clf,
imagesc(xg,yg,F1);      
colormap(flipud(pink))
axis xy
colorbar
hold on

keyboard

function iout = Ineval(zg)
iout = zeros(numel(zg),1);
iout(abs(zg) > 1) = 1;
end
